### PR TITLE
[SPARK-24373][SQL] Add AnalysisBarrier to RelationalGroupedDataset's and KeyValueGroupedDataset's child

### PR DESCRIPTION
**Pull request by mgaido91**
## What changes were proposed in this pull request?

When we create a `RelationalGroupedDataset` or a `KeyValueGroupedDataset` we set its child to the `logicalPlan` of the `DataFrame` we need to aggregate. Since the `logicalPlan` is already analyzed, we should not analyze it again. But this happens when the new plan of the aggregate is analyzed.

The current behavior in most of the cases is likely to produce no harm, but in other cases re-analyzing an analyzed plan can change it, since the analysis is not idempotent. This can cause issues like the one described in the JIRA (missing to find a cached plan).

The PR adds an `AnalysisBarrier` to the `logicalPlan` which is used as child of `RelationalGroupedDataset` or a `KeyValueGroupedDataset`.

## How was this patch tested?

added UT

### CODE REVIEW

**Comment by TomaszGaweda** [comment]
Probably similar change should be also in KeyValueGroupedDataset. It also uses logicalPlan without AnalysisBarrie

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala [hunk 1]
```diff
@@ -63,17 +63,17 @@ class RelationalGroupedDataset protected[sql](
...
+          df.sparkSession, Aggregate(Seq(Rollup(groupingExprs)), aliasedAgg, df.planWithBarrier))
       case RelationalGroupedDataset.CubeType =>
         Dataset.ofRows(
-          df.sparkSession, Aggregate(Seq(Cube(groupingExprs)), aliasedAgg, df.logicalPlan))
+          df.sparkSession, Aggregate(Seq(Cube(groupingExprs)), aliasedAgg, df.planWithBarrier))
       case RelationalGroupedDataset.PivotType(pivotCol, values) =>
         val aliasedGrps = groupingExprs.map(alias)
         Dataset.ofRows(
-          df.sparkSession, Pivot(Some(aliasedGrps), pivotCol, values, aggExprs, df.logicalPlan))
+          df.sparkSession, Pivot(Some(aliasedGrps), pivotCol, values, aggExprs, df.planWithBarrier))
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: The fix is not complete. For example, `flatMapGroupsInPandas` in RelationalGroupedDataset.scala

**Comment by gatorsmile** [comment]
@TomaszGaweda Yes, we should also fix `KeyValueGroupedDataset `.

**Comment by gatorsmile** [comment]
cc @mgaido91 Could you add a test suite to check all the external APIs in `KeyValueGroupedDataset ` and `RelationalGroupedDataset` are covered by the fix?

**Comment by mgaido91** [comment]
retest this please

File: sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+import org.apache.spark.sql.test.SharedSQLContext
+import org.apache.spark.sql.types.{LongType, StructField, StructType}
+
+class GroupedDatasetSuite extends QueryTest with SharedSQLContext {
+  import testImplicits._
+
+  private val scalaUDF = udf((x: Long) => { x + 1 })
+  private lazy val datasetWithUDF = spark.range(1).toDF("s").select($"s", scalaUDF($"s"))
+
+  private def assertContainsAnalysisBarrier(ds: Dataset[_], atLevel: Int = 1): Assertion = {
```
Comment (viirya) [inline sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala]: `withinLevel`? From the test code, seems it doesn't expect a barrier exactly at the level but within the level?

**Comment by viirya** [comment]
@mgaido91 Thanks for fixing this.

File: sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala (hunk 2 above)
Comment (mgaido91) [reply sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala]: Indeed, it expects a barrier exactly at level `atLevel` and not within. May you please recheck the code?

**Comment by viirya** [comment]
Btw, let's also mention `KeyValueGroupedDataset` in the title and description?

**Comment by mgaido91** [comment]
thanks @viirya, I forgot. Now it is updated, thanks.

File: sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala (hunk 2 above)
Comment (viirya) [reply sql/core/src/test/scala/org/apache/spark/sql/GroupedDatasetSuite.scala]: nit: Do we need to return an `Assertion` here?

**Comment by cloud-fan** [comment]
thanks, merging to master/2.3!

File: sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala (hunk 1 above)
Comment (WenboZhao) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: Any plan to make this fix complete?

Comment (mgaido91) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: It was already completed.

Comment (WenboZhao) [reply sql/core/src/main/scala/org/apache/spark/sql/RelationalGroupedDataset.scala]: ha~ Thanks!

**Pull request closed without merging**